### PR TITLE
[JBPM-9799] - Create auto-generated PIM DDL scripts by Hibernate

### PR DESCRIPTION
**Pull request by afalhambra**
**[JBPM-9799](https://issues.redhat.com/browse/JBPM-9799)**: Create auto-generated PIM DDL scripts by Hibernate

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>a pull request</b> please add comment: <b>Jenkins retest this</b>
* <b>a full downstream build</b> please add comment: <b>Jenkins run fdb</b>

</details>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -265,19 +275,43 @@ By default, the user properties IdentityProvider is provided in plain text:
 
 ## Configuring Quartz
 
-By default Quartz jobs are stored in-memory but it is possible to configure Quartz to persist the jobs in
-a database by defining the `quarkus.quartz.store-type` to either `jdbc-cmt` or `jdbc-tx`. See the 
+By default, Quartz jobs are stored in-memory database through connections managed by container, but it is possible to configure Quartz to persist the jobs in
```
Comment (ruromero) [inline README.md]:
```suggestion
By default, Quartz jobs are stored in-memory database through connections managed by the container, but it is possible to configure Quartz to persist the jobs in
```

File: README.md [hunk 2]
```diff
@@ -265,19 +275,43 @@ By default, the user properties IdentityProvider is provided in plain text:
...
+a different way either using a user transaction based configuration or the in-memory database. This can be achieved by setting the `quarkus.quartz.store-type` property 
+to `ram` or `jdbc-tx`. See the 
 [examples](./examples/quartz) and the [Quarkus Quartz documentation](https://quarkus.io/guides/quartz).
 
 ## Using other JDBC extensions
 
 The H2 JDBC extension is set by default. However, users will be able to use different JDBC extensions to connect to any
 supported database. For that purpose you will have to re-augment the base build to include the right build properties.
 
+For instance, below command will add a PostgreSQL database extension by re-augmenting the application.
```
Comment (ruromero) [inline README.md]:
```suggestion
For instance, the below command will add a PostgreSQL database extension by re-augmenting the application.
```

File: README.md [hunk 3]
```diff
@@ -265,19 +275,43 @@ By default, the user properties IdentityProvider is provided in plain text:
...
 
-* [Quarkus Datasources](https://quarkus.io/guides/datasource#jdbc-datasource).
+* [Quarkus Datasources](https://quarkus.io/guides/datasource#jdbc-datasource)
 * [Quarkus Re-augmentation](https://quarkus.io/guides/reaugmentation)
+* [Quarkus Flyway](https://quarkus.io/guides/flyway)
+
+## Disabling database schema auto-creation
+
+Process Migration service generates the database schema automatically if it is not already present by using the DDL scripts bundled in the application.
+You should be able to disable and manage the schema creation by your own by disabling some Flyway properties.
```
Comment (ruromero) [inline README.md]:
```suggestion
You should be able to disable and manage the schema creation by disabling some Flyway properties.
```

File: ddl-scripts/db2/db2-pim-schema.sql [hunk 4]
```diff
@@ -0,0 +1,80 @@
+create sequence MIG_REP_ID_SEQ start with 1 increment by 50;
+create sequence MIGRATION_ID_SEQ start with 1 increment by 50;
+create sequence PLAN_ID_SEQ start with 1 increment by 50;
+
+create table migration_report_logs (
+   report_id bigint not null,
```
Comment (ruromero) [inline ddl-scripts/db2/db2-pim-schema.sql]: the first line of each `create` seems to be one space short

File: ddl-scripts/db2/db2-pim-schema.sql [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    target_process_id varchar(255),
+    primary key (id)
+);
+
+create table process_instance_ids (
+   migration_definition_id bigint not null,
+    processInstanceIds bigint
+);
+
+create index IDXgiy3fyawbd9nt7mymgd9qd61h on migration_reports (migration_id);
```
Comment (ruromero) [inline ddl-scripts/db2/db2-pim-schema.sql]: Can we use a more descriptive name for this index?

File: ddl-scripts/h2/h2-pim-schema.sql [hunk 6]
```diff
@@ -0,0 +1,80 @@
+create sequence MIG_REP_ID_SEQ start with 1 increment by 50;
+create sequence MIGRATION_ID_SEQ start with 1 increment by 50;
+create sequence PLAN_ID_SEQ start with 1 increment by 50;
+
+create table migration_report_logs (
+   report_id bigint not null,
```
Comment (ruromero) [inline ddl-scripts/h2/h2-pim-schema.sql]: same, please fix the indentation in this file too

File: ddl-scripts/h2/h2-pim-schema.sql [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    target_process_id varchar(255),
+    primary key (id)
+);
+
+create table process_instance_ids (
+   migration_definition_id bigint not null,
+    processInstanceIds bigint
+);
+
+create index IDXgiy3fyawbd9nt7mymgd9qd61h on migration_reports (migration_id);
```
Comment (ruromero) [inline ddl-scripts/h2/h2-pim-schema.sql]: index name

File: ddl-scripts/mariadb/mariadb-pim-schema.sql [hunk 8]
```diff
@@ -0,0 +1,80 @@
+create sequence MIG_REP_ID_SEQ start with 1 increment by 50;
```
Comment (ruromero) [inline ddl-scripts/mariadb/mariadb-pim-schema.sql:1]: indentation and index name

File: src/test/java/org/kie/processmigration/persistence/GenerateDDLScriptsTest.java [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+        public Path buildDropFile(Path basePath) {
+            return basePath.resolve(alias).resolve(prefix + "-pim-drop-schema.sql");
+
+        }
+
+        public String getDialect() {
+            return this.dialect;
+        }
+
+        @Override
```
Comment (ruromero) [inline src/test/java/org/kie/processmigration/persistence/GenerateDDLScriptsTest.java]: You can remove these methods and use lombok that is already included

File: src/test/java/org/kie/processmigration/service/PlanServiceImplTest.java [hunk 10]
```diff
@@ -43,7 +42,7 @@
     @Inject
     PlanService planService;
 
-    @AfterEach
+    @BeforeEach
```
Comment (ruromero) [inline src/test/java/org/kie/processmigration/service/PlanServiceImplTest.java]: Why is this needed? I'd clean up something after I've done changes

File: src/test/java/org/kie/processmigration/service/SchedulerServiceImplTest.java [hunk 11]
```diff
@@ -92,9 +90,13 @@ void testScheduleMigrationNow() throws InvalidMigrationException, InterruptedExc
...
 
         // Then
-        count.await(10, TimeUnit.SECONDS);
+        if (!count.await(10, TimeUnit.SECONDS)) {
+            fail("Failed while waiting for the jobs to be completed");
+        }
         verify(migrationService, times(1)).get(99L);
         verify(migrationService, times(1)).migrate(migration);
+
+        Thread.sleep(1000L); // wait for the triggers/jobs to be deleted from the db
```
Comment (ruromero) [inline src/test/java/org/kie/processmigration/service/SchedulerServiceImplTest.java]: Can't we avoid using `Thread.sleep()` ?

**Review by ruromero — CHANGES_REQUESTED** [review]
Hey @afalhambra 
Great PR! Just minor changes and small comments. I like your approach for using Flyway.
I know scripts are generated and making changes to them manually might be an unnecessary extra effort so disregard these comments if you want.

File: README.md (hunk 1 above)
Comment (afalhambra) [reply README.md]: done

File: README.md (hunk 2 above)
Comment (afalhambra) [reply README.md]: done

File: README.md (hunk 3 above)
Comment (afalhambra) [reply README.md]: done

File: ddl-scripts/db2/db2-pim-schema.sql (hunk 4 above)
Comment (afalhambra) [reply ddl-scripts/db2/db2-pim-schema.sql]: done

File: ddl-scripts/db2/db2-pim-schema.sql (hunk 5 above)
Comment (afalhambra) [reply ddl-scripts/db2/db2-pim-schema.sql]: yes, thanks. Done

File: src/test/java/org/kie/processmigration/persistence/GenerateDDLScriptsTest.java (hunk 9 above)
Comment (afalhambra) [reply src/test/java/org/kie/processmigration/persistence/GenerateDDLScriptsTest.java]: yes, thanks for that

File: src/test/java/org/kie/processmigration/service/PlanServiceImplTest.java (hunk 10 above)
Comment (afalhambra) [reply src/test/java/org/kie/processmigration/service/PlanServiceImplTest.java]: Originally the problem was that flyway didn't clean up the database, so there might be some leftover from previous tests. I now forced flyway to clean the db at startup for the tests, so it should be ok now.

File: src/test/java/org/kie/processmigration/service/SchedulerServiceImplTest.java (hunk 11 above)
Comment (afalhambra) [reply src/test/java/org/kie/processmigration/service/SchedulerServiceImplTest.java]: removing last checks, since they are not part of the goal of these tests. Also, seems there is some problems with the listeners and the checkExists method retrieve the job even though the job has been already completed.

**Review by ruromero — APPROVED** [review]
Great!

**Comment by afalhambra** [comment]
@mareknovotny @Ginxo could you please have a look at below CI error?
```
[Pipeline] sh
[2021-09-07T12:45:11.613Z] Warning: A secret was passed to "sh" using Groovy String interpolation, which is insecure.
[2021-09-07T12:45:11.613Z] 		 Affected argument(s) used the following variable(s): [GITHUB_TOKEN]
[2021-09-07T12:45:11.613Z] 		 See https://jenkins.io/redirect/groovy-string-interpolation for details.
[2021-09-07T12:45:12.060Z] + build-chain-action -token=**** '-df=https://raw.githubusercontent.com/${GROUP}/droolsjbpm-build-bootstrap/${BRANCH}/.ci/pull-request-config.yaml' -folder=bc build pr -url=https://github.com/kiegroup/process-migration-service/pull/11 --skipParallelCheckout -cct '(^mvn .*)||$1 -s /home/jenkins/workspace/KIE/main/pullrequest/process-migration-service-main.pr@tmp/config9199728961119115548tmp -Dmaven.wagon.http.ssl.insecure=true'
[2021-09-07T12:45:13.501Z] [INFO]  Executing pull request flow for https://github.com/kiegroup/process-migration-service/pull/11 in bc
[2021-09-07T12:45:14.453Z] ::group::[Pull Request Flow] Checking out kiegroup/process-migration-service and its dependencies
[2021-09-07T12:45:14.453Z] [ERROR]  Cannot read property 'parents' of undefined
```

**Comment by mareknovotny** [comment]
jenkins retest this please

**Comment by mareknovotny** [comment]
@afalhambra the failures in new tests ;)

- org.kie.processmigration.persistence.DDLScriptsTest.testValidateMigration{Execution}[1]
- org.kie.processmigration.persistence.DDLScriptsTest.testValidateMigration{Execution}[2]
